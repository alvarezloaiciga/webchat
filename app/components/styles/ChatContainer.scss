$chatContainerBottomRight: 24px;

.ChatContainer {
  &.hidden {
    display: none;
  }

  &.standaloneMode {
    width: 100vw !important;
    height: 100vh !important;
    max-width: none !important;
    max-height: none !important;
    margin: 0;
    padding: 0;
    position: initial;
    right: 0;
    bottom: 0;
    border: none;
    border-radius: 0;
    animation: none;
  }

  // width and max-height set with inline style through config
  height: calc(100vh - 132px);
  position: fixed;
  right: $chatContainerBottomRight;
  bottom: 108px;
  border-radius: 5px;
  overflow: hidden;
  box-shadow: rgba(0, 0, 0, 0.117647) 0px 1px 6px, rgba(0, 0, 0, 0.117647) 0px 1px 4px;
  animation: 0.3s chatDialogOpen 1;
  display: flex;
  flex-direction: column;
  background: #f4f4f8;

  &.hasCustomLauncher {
    bottom: $chatContainerBottomRight;
  }

  .chatContainerBody {
    flex: 1 1 auto;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    overflow: hidden;
  }

  .banner,
  .errorBanner {
    flex: 0 0 auto;
    color: #fff;
    display: flex;
    flex-direction: column;
    justify-content: center;
    font-weight: 300;
    font-size: 14px;
    padding: 0 30px;
    align-items: stretch;
    text-align: center;
  }

  .banner {
    background: #59ad5d;
    height: 60px;
  }

  .errorBanner {
    background: #ad2215;
    height: 50px;
  }
}

.noFlexbox .ChatContainer {
  // Default demensions
  width: 400px;
  height: 600px;

  .banner,
  .errorBanner {
    width: 342px;
    max-width: 342px;
    min-width: 342px;
    text-align: center;
  }

  .banner,
  .errorBanner {
    position: absolute;
    top: 30px;
    height: 50px;
    min-height: 50px;
    max-height: 50px;
    display: table;

    span {
      display: table-cell;
      vertical-align: middle;
    }
  }
}
