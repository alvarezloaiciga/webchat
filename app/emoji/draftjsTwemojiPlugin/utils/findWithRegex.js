const findWithRegex = (regex, contentBlock, callback) => {
  // Get the text from the contentBlock
  const text = contentBlock.getText();
  let matchArr;
  let start; // eslint-disable-line
  // Go through all matches in the text and return the indices to the callback
  // eslint-disable-next-line no-cond-assign
  while ((matchArr = regex.exec(text)) !== null) {
    start = matchArr.index;
    callback(start, start + matchArr[0].length);
  }
};

export default findWithRegex;
