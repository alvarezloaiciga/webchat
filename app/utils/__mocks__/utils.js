export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const isIEorSafari = () => false;
