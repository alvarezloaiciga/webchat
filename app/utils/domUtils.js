import ReactDOM, {render} from 'react-dom';
import React from 'react';
import Routes from 'Routes';
import {quiqContainerId} from 'Common/Constants';
import {AppContainer} from 'react-hot-loader';
import {IntlProvider} from 'react-intl';
import Redbox from 'redbox-react';
import {Provider} from 'react-redux';

export const appIsMounted = (): boolean => {
  const container = document.getElementById(quiqContainerId);
  if (!container) return false;
  return container.childElementCount > 0;
};

export const destructApp = () => {
  ReactDOM.unmountComponentAtNode(document.getElementById(quiqContainerId));
};

export const constructApp = store => {
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <AppContainer errorReporter={Redbox}>
          <Routes />
        </AppContainer>
      </IntlProvider>
    </Provider>,
    document.getElementById(quiqContainerId),
  );

  if (module.hot) {
    module.hot.accept('Routes', () => {
      const NextApp = require('Routes').default; // eslint-disable-line global-require

      render(
        <Provider store={store}>
          <IntlProvider locale="en">
            <AppContainer errorReporter={Redbox}>
              <NextApp />
            </AppContainer>
          </IntlProvider>
        </Provider>,
        document.getElementById(quiqContainerId),
      );
    });
  }
};

export const removeElement = (e: HTMLElement) => {
  e.parentNode.removeChild(e);
};

export const reloadApp = store => {
  destructApp();
  constructApp(store);
};

export const setFavicon = (iconUrl: string, appleTouchUrl: ?string) => {
  // Remove all icons
  const icons = [
    ...Array.from(document.querySelectorAll("link[rel='icon']")),
    ...Array.from(document.querySelectorAll("link[rel='apple-touch-icon']")),
  ];
  icons.forEach(icon => {
    removeElement(icon);
  });

  // Build icon
  const icon = document.createElement('link');
  icon.rel = 'icon';
  icon.href = iconUrl;

  // Build apple touch icon
  const appleTouchIcon = document.createElement('link');
  appleTouchIcon.rel = 'apple-touch-icon';
  appleTouchIcon.href = appleTouchUrl || iconUrl;

  // NOTE HARD: The apple-touch-icon must come BEFORE the regular icon.
  // Chrome looks for the last meta tag with a rel containing "icon" to try and use as title bar icon.
  const head = document.getElementsByTagName('head')[0];
  head.appendChild(appleTouchIcon);
  head.appendChild(icon);
};

export const setApplicationName = (name: string) => {
  const appName = document.querySelector("meta[name='application-name']");
  if (appName) {
    removeElement(appName);
  }
  const newAppName = document.createElement('meta');
  newAppName.name = 'application-name';
  newAppName.content = name;
  document.getElementsByTagName('head')[0].appendChild(newAppName);
};

export const setBrowserThemeColor = (color: string) => {
  const colorTag = document.querySelector("meta[name='theme-color']");
  if (colorTag) {
    removeElement(colorTag);
  }
  const newColor = document.createElement('meta');
  newColor.name = 'theme-color';
  newColor.content = color;
  document.getElementsByTagName('head')[0].appendChild(newColor);
};
