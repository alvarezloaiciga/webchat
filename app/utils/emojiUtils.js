import {emojiIndex} from 'emoji-mart';
import emojiRegexFactory from 'emoji-regex';
import invertBy from 'lodash/invertBy';
import type {EmojiMetadata, Emoji} from 'Common/Types';
import quiqOptions from 'Common/QuiqOptions';

const emojiRegex = emojiRegexFactory();

// Build lookup table of {unifiedCode : emojiId}
const emojiIdByUnified = {};
Object.keys(emojiIndex.emojis).forEach(id => {
  const unified = emojiIndex.emojis[id].unified.toLowerCase();
  emojiIdByUnified[unified] = id;
});

export const isSingleEmoji = (text: string): boolean => {
  // We know a string contains only one emoji iff
  // 1) An emoji regex finds only one match in the string, and
  // 2) The number of characters in that emoji === the number of characters in the string
  const emojis = text.match(emojiRegex);

  if (!emojis)
    return false;

  const emojiCharCount = emojis.reduce((count, e) => count + e.length, 0);
  return emojis.length === 1 && text.length === emojiCharCount;
};

export const convertUnicodeToEmojiObject = (u: string): ?Emoji => {
  const emojiId = Object.keys(emojiIndex.emojis).find(k => emojiIndex.emojis[k].native === u);
  return emojiIndex.emojis[emojiId];
};

// This function is used for filtering emojis in our picker, as well as redacting emojis prior to sending a message.
export const emojiFilter = (e: EmojiMetadata | Emoji | string) => {
  let emojiId;
  // emoji-mart will call this function when user is searching, but it only passes unicode string:
  if (typeof e === 'string') {
    emojiId =  emojiIdByUnified[e.toLowerCase()];
  } else {
    // NOTE: In emoji-mart land, id === short_codes[0]
    emojiId = e.id || e.short_names[0];
  }

  // Don't include any emojis that don't have id or shortcode (we don't know what they are)
  if (!emojiId)
    return false;

  const {includeEmojis, excludeEmojis} = quiqOptions;
  // Prioritize include over exclude. Only one of the should be defined by customer, however.
  if (Array.isArray(includeEmojis)) {
    return includeEmojis.includes(emojiId);
  }
  else if (Array.isArray(excludeEmojis)) {
    return !excludeEmojis.includes(emojiId);
  }

  // If neither exclude nor include was defined, always include emoji.
  return true;
};
