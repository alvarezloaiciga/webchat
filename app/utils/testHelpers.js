// @flow
import type {Message, IntlObject, IntlMessage} from 'types';

export const getMockMessage = (
  // eslint-disable-line import/prefer-default-export
  id?: number = 0,
  overrides?: Message | {} = {},
) => {
  const message = {
    authorType: id % 2 === 0 ? 'Customer' : 'Agent',
    text: `text--${id}`,
    id: `id--${id}`,
    timestamp: id,
    type: 'Text',
  };

  return Object.assign(message, overrides);
};

/**
 * Stub out formatMessage so that it will return the defaultMessage with basic variable replacement
 */
const formatTestMessage = (msg: IntlMessage, values?: {[key: string]: string}) => {
  if (!values) {
    return msg.defaultMessage;
  }

  let returnValue = msg.defaultMessage;
  Object.keys(values).forEach(key => {
    if (values) {
      const value = values[key];
      returnValue = returnValue.replace(`{${key}}`, value);
    }
  });

  return returnValue;
};

export const TestIntlObject: IntlObject = {
  formatMessage: formatTestMessage,
  formatRelative: value => value.toString(),
  formatTime: timestamp => timestamp.toString(),
  formatDate: timestamp => timestamp.toString(),
};
