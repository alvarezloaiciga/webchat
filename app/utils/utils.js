// @flow
declare var Modernizr: Object;
import {UAParser} from 'ua-parser-js';
import type {BrowserNames, DeviceTypes, OSNames} from 'types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// Modernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore Modernizr in that case
export const supportsFlexbox = () => isIE10() || Modernizr.flexbox;
