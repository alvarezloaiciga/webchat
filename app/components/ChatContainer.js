// @flow
import React from 'react';
import {getStyle} from 'Common/QuiqOptions';
import {
  inStandaloneMode,
  isStorageEnabled,
  isSupportedBrowser,
  uuidv4,
  convertToExtensionMessages,
  isIPhone,
} from 'Common/Utils';
import {createGuid} from 'core-ui/utils/stringUtils';
import WelcomeForm from 'WelcomeForm';
import MessageForm from 'MessageForm';
import Debugger from './Debugger/Debugger';
import HeaderMenu from 'HeaderMenu';
import QuiqChatClient from 'quiq-chat';
import Transcript from 'Transcript';
import Spinner from 'Spinner';
import {connect} from 'react-redux';
import {
  ChatInitializedState,
  intlMessageTypes,
  maxAttachmentSize,
  ExtensionSdkEventTypes,
  AttachmentErrorTypes,
} from 'Common/Constants';
import Dropzone from 'react-dropzone';
import * as ChatActions from 'actions/chatActions';
import type {
  ChatState,
  ChatInitializedStateType,
  ChatConfiguration,
  Message as MessageType,
  AttachmentError,
} from 'Common/types';
import {registerExtension, postExtensionEvent} from 'services/Extensions';
import {
  getTranscript,
  getIsAgentAssigned,
  getAgentEndedLatestConversation,
  getConfiguration,
  getMessage,
} from 'reducers/chat';
import styled, {css} from 'react-emotion';

export const getHeight = (newHeight: string, heightOverride: number): string => {
  let height = newHeight;

  if (isIPhone()) {
    height = `${heightOverride}px`;
  }

  return height;
};

export const ChatContainerStyle = styled.div`
  width: 99vw !important;
  height: ${props => props.height} !important;
  max-width: none !important;
  max-height: none !important;
  margin: auto !important;
  padding: 0;
  border: 1px solid rgba(0, 0, 0, 0.117647);
  border-radius: 5px;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  background: #f4f4f8;

  ${props =>
    props.standaloneMode &&
    css`
      width: 100vw !important;
      position: initial;
      right: 0;
      bottom: 0;
      border: none;
      border-radius: 0;
      animation: none;
    `};
`;

export const ChatContainerBody = styled.div`
  flex: 1 1 auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  overflow: hidden;
`;

export const Banner = styled.div`
  flex: 0 0 auto;
  color: #fff;
  display: flex;
  flex-direction: column;
  justify-content: center;
  font-weight: 300;
  font-size: 14px;
  padding: 0 30px;
  align-items: stretch;
  text-align: center;
  background: #59ad5d;
  height: 60px;
`;

export const ErrorBanner = styled.div`
  flex: 0 0 auto;
  color: #fff;
  display: flex;
  flex-direction: column;
  justify-content: center;
  font-weight: 300;
  font-size: 14px;
  padding: 0 30px;
  align-items: stretch;
  text-align: center;
  background: #ad2215;
  height: 50px;
`;

/* eslint-disable no-confusing-arrow */
export const TranscriptArea = styled(Dropzone)`
  display: flex;
  flex: ${props => (props.haswaitscreen === 'true' ? 0 : 1)} 1 auto;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  width: 100%;
  white-space: pre-wrap;
`;
/* eslint-disable no-confusing-arrow */

export const WaitScreenScrollWrapper = styled.div`
  flex: 1 1 auto;
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  // IMPORTANT: This property is needed to allow for scrolling within the iFrame
  // on mobile devices. If you remove, be sure to test those scenarios.
  -webkit-overflow-scrolling: touch;
`;

export const WaitScreen = styled.iframe`
  flex: 1 1 auto;
  height: 100%;
  border-width: 0;
  width: 100%;
`;

export type ChatContainerProps = {
  chatContainerHidden: boolean,
  configuration: ChatConfiguration,
  welcomeFormRegistered: boolean,
  initializedState: ChatInitializedStateType,
  isAgentAssigned: boolean,
  transcript: Array<MessageType>,
  agentEndedConversation: boolean,
  agentsAvailable: boolean,
  setUploadProgress: (messageId: string, progress: number) => void,
  updatePendingAttachmentId: (tempId: string, newId: string) => void,
  addPendingAttachmentMessage: (
    tempId: string,
    contentType: string,
    url: string,
    fromCustomer: boolean,
  ) => void,
  addAttachmentError: (attachmentError: AttachmentError) => void,
  removeMessage: (id: string) => void,
};

export type ChatContainerState = {
  bannerMessage?: string,
  agentsAvailableOrSubscribed: boolean,
  heightOverride: number,
};

export class ChatContainer extends React.Component<ChatContainerProps, ChatContainerState> {
  props: ChatContainerProps;
  state: ChatContainerState = {
    agentsAvailableOrSubscribed: false,
    heightOverride: 0,
  };
  dropzone: ?Dropzone;
  bannerMessageTimeout: ?number;
  extensionFrame: any;

  componentWillMount() {
    // Set custom window title
    document.title = getMessage(intlMessageTypes.pageTitle);

    this.setState({
      agentsAvailableOrSubscribed: this.props.agentsAvailable || QuiqChatClient.isUserSubscribed(),
      // Save off the original inner height, on IOS 10.3, this height can change
      // when the keyboard is displayed, which puts us in a bad state. If you remove
      // this be sure to test using the keyboard and return key in IOS 10.3.
      heightOverride:
        window.innerHeight > window.innerWidth ? window.innerHeight : window.innerWidth,
    });
  }

  componentWillReceiveProps(nextProps: ChatContainerProps) {
    this.setState({
      agentsAvailableOrSubscribed: nextProps.agentsAvailable || QuiqChatClient.isUserSubscribed(),
    });
  }

  displayTemporaryError = (text: string, duration: number) => {
    // Clear any pending timeout
    if (this.bannerMessageTimeout) {
      clearTimeout(this.bannerMessageTimeout);
    }

    this.setState({
      bannerMessage: text,
    });

    // Hide the error in <duration> milliseconds
    this.bannerMessageTimeout = setTimeout(
      () => this.setState({bannerMessage: undefined}),
      duration,
    );
  };

  handleAttachments = (accepted: Array<File>, rejected: Array<File>) => {
    rejected.forEach(f => {
      this.props.addAttachmentError({
        timestamp: Date.now(),
        id: createGuid(),
        type:
          f.size > maxAttachmentSize
            ? AttachmentErrorTypes.TOO_LARGE
            : AttachmentErrorTypes.UNSUPPORTED_TYPE,
        payload: f.name,
      });
    });

    if (rejected.length > 0) return;

    // Clear any attachment error
    this.setState({bannerMessage: undefined});

    accepted.forEach(file => {
      const tempId = `temp_${uuidv4()}`;
      const dataUrl = window.URL.createObjectURL(file);
      this.props.addPendingAttachmentMessage(tempId, file.type, dataUrl, true);
      QuiqChatClient.sendAttachmentMessage(file, (progress: number) =>
        this.props.setUploadProgress(tempId, progress),
      )
        .then(id => {
          this.props.updatePendingAttachmentId(tempId, id);
        })
        .catch(() => {
          this.props.addAttachmentError({
            timestamp: Date.now(),
            id: createGuid(),
            type: AttachmentErrorTypes.UPLOAD_ERROR,
            payload: file.name,
          });
          this.props.removeMessage(tempId);
        });
    });
  };

  openFileBrowser = () => {
    if (this.dropzone) {
      this.dropzone.open();
    }
  };

  renderBanner = () => {
    const {colors, styles, fontFamily} = this.props.configuration;

    const bannerStyle = getStyle(styles.HeaderBanner, {
      backgroundColor: colors.primary,
      fontFamily,
    });

    const errorBannerStyle = getStyle(styles.ErrorBanner, {fontFamily});

    // If state indicates a warning message, use that
    if (this.state.bannerMessage) {
      return <ErrorBanner style={errorBannerStyle}>{this.state.bannerMessage}</ErrorBanner>;
    }

    switch (this.props.initializedState) {
      case ChatInitializedState.INITIALIZED:
      case ChatInitializedState.LOADING:
      case ChatInitializedState.UNINITIALIZED:
        return <Banner style={bannerStyle}>{getMessage(intlMessageTypes.headerText)}</Banner>;
      case ChatInitializedState.INACTIVE:
        return (
          <ErrorBanner style={errorBannerStyle}>
            {getMessage(intlMessageTypes.inactiveMessage)}
          </ErrorBanner>
        );
      case ChatInitializedState.DISCONNECTED:
        return (
          <ErrorBanner style={errorBannerStyle}>
            {getMessage(intlMessageTypes.reconnectingMessage)}
          </ErrorBanner>
        );
      case ChatInitializedState.ERROR:
      case ChatInitializedState.BURNED:
      default:
        return (
          <ErrorBanner style={errorBannerStyle}>
            {getMessage(intlMessageTypes.errorMessage)}
          </ErrorBanner>
        );
    }
  };

  renderContent = () => {
    const chatContainerStyle = {
      backgroundColor: this.props.configuration.colors.transcriptBackground,
    };
    switch (this.props.initializedState) {
      case ChatInitializedState.INITIALIZED:
        return (
          <ChatContainerBody style={chatContainerStyle}>
            {this.isUsingWaitScreen() && (
              // IMPORTANT: This wrapper is needed to get scrolling and the flex resizing to
              // working correctly on mobile devices. If you remove, be sure to test those
              // scenarios.
              <WaitScreenScrollWrapper
                style={{
                  minHeight: this.getWaitScreenMinHeight(),
                  height: this.getWaitScreenHeight(),
                  borderWidth: 0,
                  flexGrow: this.getWaitScreenFlexGrow(),
                  width: '100%',
                }}
              >
                <WaitScreen
                  innerRef={r => {
                    this.extensionFrame = r;
                  }}
                  style={{
                    minHeight: this.getWaitScreenMinHeight(),
                  }}
                  onLoad={this.handleIFrameLoad}
                  sandbox="allow-scripts allow-popups allow-popups-to-escape-sandbox allow-forms allow-same-origin allow-orientation-lock"
                  src={
                    this.props.configuration.customScreens &&
                    this.props.configuration.customScreens.waitScreen
                      ? this.props.configuration.customScreens.waitScreen.url
                      : ''
                  }
                />
              </WaitScreenScrollWrapper>
            )}
            <TranscriptArea
              innerRef={d => {
                this.dropzone = d;
              }}
              haswaitscreen={this.isUsingWaitScreen().toString()}
              disabled={
                !this.state.agentsAvailableOrSubscribed ||
                !this.props.configuration.enableChatFileAttachments ||
                (this.props.configuration.enableManualConvoStart &&
                  this.props.agentEndedConversation)
              }
              accept={this.props.configuration.supportedAttachmentTypes.join(',')}
              disablePreview={true}
              disableClick={true}
              maxSize={maxAttachmentSize}
              onDrop={this.handleAttachments}
              style={{
                // This is to ensure that the size of this renders in a way that allows us to at least scroll in IE 10
                minHeight: this.props.transcript.length > 0 ? '75px' : '0px',
              }}
              id="TranscriptArea"
            >
              <Transcript />
              <MessageForm openFileBrowser={this.openFileBrowser} />
            </TranscriptArea>
          </ChatContainerBody>
        );
      case ChatInitializedState.UNINITIALIZED:
      case ChatInitializedState.LOADING:
        return (
          <ChatContainerBody style={chatContainerStyle}>
            <Spinner />
          </ChatContainerBody>
        );
      case ChatInitializedState.DISCONNECTED:
      case ChatInitializedState.ERROR:
      case ChatInitializedState.INACTIVE:
      case ChatInitializedState.BURNED:
      default:
        return (
          <ChatContainerBody style={chatContainerStyle}>
            <Transcript />
          </ChatContainerBody>
        );
    }
  };

  isUsingWaitScreen = () => {
    return !!(
      this.props.configuration.customScreens &&
      this.props.configuration.customScreens.waitScreen &&
      !this.props.isAgentAssigned &&
      !this.props.agentEndedConversation
    );
  };

  handleIFrameLoad = () => {
    // $FlowIssue - Null check is done upstream of this call
    registerExtension(
      // $FlowIssue - Null check is done upstream of this call
      this.props.configuration.customScreens.waitScreen.url,
      this.extensionFrame.contentWindow,
    );

    postExtensionEvent({
      eventType: 'transcriptChanged',
      data: {messages: convertToExtensionMessages(this.props.transcript)},
    });

    postExtensionEvent({
      eventType: ExtensionSdkEventTypes.ESTIMATED_WAIT_TIME_CHANGED,
      data: {estimatedWaitTime: QuiqChatClient.getEstimatedWaitTime()},
    });

    QuiqChatClient.onEstimatedWaitTimeChanged((estimatedWaitTime?: number) => {
      if (this.extensionFrame && this.extensionFrame.contentWindow) {
        postExtensionEvent({
          eventType: ExtensionSdkEventTypes.ESTIMATED_WAIT_TIME_CHANGED,
          data: {estimatedWaitTime},
        });
      }
    });
  };

  getWaitScreenHeight = () => {
    // $FlowIssue - Null check is done upstream of this call
    return this.props.configuration.customScreens.waitScreen.height
      ? this.props.configuration.customScreens.waitScreen.height
      : '100%';
  };

  getWaitScreenFlexGrow = () => {
    // $FlowIssue - Null check is done upstream of this call
    return this.props.configuration.customScreens.waitScreen.height ? 0 : 1;
  };

  getWaitScreenMinHeight = () => {
    // $FlowIssue - Null check is done upstream of this call
    return this.props.configuration.customScreens.waitScreen.minHeight
      ? this.props.configuration.customScreens.waitScreen.minHeight
      : 100;
  };

  render() {
    if (this.props.chatContainerHidden || !isSupportedBrowser() || !isStorageEnabled()) return null;

    const height = inStandaloneMode() ? '100vh' : '98vh';

    if (
      this.props.configuration.demoMode ||
      (this.props.initializedState === ChatInitializedState.INITIALIZED &&
        !this.props.welcomeFormRegistered)
    ) {
      return (
        <ChatContainerStyle
          standaloneMode={inStandaloneMode()}
          height={getHeight(height, this.state.heightOverride)}
        >
          <WelcomeForm />
        </ChatContainerStyle>
      );
    }

    return (
      <ChatContainerStyle
        standaloneMode={inStandaloneMode()}
        height={getHeight(height, this.state.heightOverride)}
      >
        <HeaderMenu />
        {this.renderBanner()}
        <Debugger />
        {this.renderContent()}
      </ChatContainerStyle>
    );
  }
}

const mapStateToProps = (state: ChatState) => ({
  chatContainerHidden: state.chatContainerHidden,
  initializedState: state.initializedState,
  welcomeFormRegistered: state.welcomeFormRegistered,
  configuration: getConfiguration(state),
  isAgentAssigned: getIsAgentAssigned(state),
  transcript: getTranscript(state),
  agentEndedConversation: getAgentEndedLatestConversation(state),
  agentsAvailable: state.agentsAvailable,
});

const mapDispatchToProps = {
  setUploadProgress: ChatActions.setUploadProgress,
  updatePendingAttachmentId: ChatActions.updatePendingAttachmentId,
  addPendingAttachmentMessage: ChatActions.addPendingAttachmentMessage,
  addAttachmentError: ChatActions.addAttachmentError,
  removeMessage: ChatActions.removeMessage,
};

export default connect(mapStateToProps, mapDispatchToProps)(ChatContainer);
