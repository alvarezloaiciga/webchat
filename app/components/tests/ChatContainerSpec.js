// @flow
import React from 'react';
import ChatContainer from '../ChatContainer';
import { getMockMessage } from 'utils/testHelpers';
import { shallow } from 'enzyme';
import type { ShallowWrapper } from 'enzyme';


describe('ChatContainer component', () => {
  let wrapper:ShallowWrapper;
  let render: () => void;

  beforeEach(() => {
    render = () => {
      wrapper = shallow(<ChatContainer />);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('loading', () => {
      describe('when not loading', () => {
        it('displays transcript', () => {
          wrapper.setState({ loading: false });
          expect(wrapper.find('Transcript').length).toBe(1);
          expect(wrapper.find('Spinner').length).toBe(0);
        });

        describe('when connected', () => {
          it('displays message form', () => {
            wrapper.setState({ loading: false, connected: true });
            expect(wrapper.find('Transcript').length).toBe(1);
            expect(wrapper.find('MessageForm').length).toBe(1);
            expect(wrapper.find('Spinner').length).toBe(0);
          });
        });
      });
    });

    describe('messages', () => {
      it('appends messages to the transcript', () => {
        wrapper.setState({
          loading: false,
          connected: true,
          messages: [getMockMessage(0), getMockMessage(1), getMockMessage(2)],
        });
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
