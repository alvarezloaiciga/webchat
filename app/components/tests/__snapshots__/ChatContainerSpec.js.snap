// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatContainer component rendering hidden hides the component when hidden 1`] = `null`;

exports[`ChatContainer component rendering messages appends messages to the transcript 1`] = `
<div
  className="ChatContainer"
>
  <div
    className="banner"
    style={
      Object {
        "backgroundColor": "#333",
      }
    }
  >
    <span
      className="messageUs"
    >
      TOOL TIME
    </span>
  </div>
  <div
    className="chatContainerBody"
  >
    <Transcript
      messages={
        Array [
          Object {
            "authorType": "Guest",
            "body": "body--0",
            "id": "id--0",
            "timestamp": 0,
            "type": "Text",
          },
          Object {
            "authorType": "Agent",
            "body": "body--1",
            "id": "id--1",
            "timestamp": 1,
            "type": "Text",
          },
          Object {
            "authorType": "Guest",
            "body": "body--2",
            "id": "id--2",
            "timestamp": 2,
            "type": "Text",
          },
        ]
      }
    />
    <MessageForm
      agentTyping={false}
    />
  </div>
</div>
`;

exports[`ChatContainer component rendering renders 1`] = `
<div
  className="ChatContainer"
>
  <div
    className="banner"
    style={
      Object {
        "backgroundColor": "#333",
      }
    }
  >
    <span
      className="messageUs"
    >
      TOOL TIME
    </span>
  </div>
  <div
    className="chatContainerBody"
  >
    <Spinner />
  </div>
</div>
`;

exports[`ChatContainer component rendering typing indicator updates typing state 1`] = `
<div
  className="ChatContainer"
>
  <div
    className="banner"
    style={
      Object {
        "backgroundColor": "#333",
      }
    }
  >
    <span
      className="messageUs"
    >
      TOOL TIME
    </span>
  </div>
  <div
    className="chatContainerBody"
  >
    <Transcript
      messages={Array []}
    />
    <MessageForm
      agentTyping={true}
    />
  </div>
</div>
`;
